Source/GTMSessionFetcherService.m

```diff
@@ -374,17 +374,19 @@ - (void)startFetcher:(GTMSessionFetcher *)fetcher {
 }
 
 // Internal utility. Returns a fetcher's delegate if it's a dispatcher, or nil if the fetcher
-// is its own delegate and has no dispatcher.
+// is its either own delegate and has no dispatcher or swizzled alien.
 - (GTMSessionFetcherSessionDelegateDispatcher *)delegateDispatcherForFetcher:(GTMSessionFetcher *)fetcher {
   GTMSessionCheckNotSynchronized(self);
 
   NSURLSession *fetcherSession = fetcher.session;
   if (fetcherSession) {
     id<NSURLSessionDelegate> fetcherDelegate = fetcherSession.delegate;
-    BOOL hasDispatcher = (fetcherDelegate != nil && fetcherDelegate != fetcher);
-    if (hasDispatcher) {
+      BOOL maybeHasDispatcher = (fetcherDelegate != nil && fetcherDelegate != fetcher);
+      
       GTMSESSION_ASSERT_DEBUG([fetcherDelegate isKindOfClass:[GTMSessionFetcherSessionDelegateDispatcher class]],
                               @"Fetcher delegate class: %@", [fetcherDelegate class]);
+      
+    if (maybeHasDispatcher && [fetcherDelegate isKindOfClass:GTMSessionFetcherSessionDelegateDispatcher.class]) {
```

Comment:
Calling it out specifically - swizzling doesn't change the type, it changes the methods on a class, so I'm not following why someone swizzling was causing problems in the existing code and this is needed instead.

Reply:
Thanks for finding a time for the MR review.
So, this is the actual problem pinpoint. 

My experiments with `TSKNSURLSessionDelegateProxy` revealed it's pretty well made wrapper for url session delegate.

What is does, in case you creating own session instance – it's wrapping your delegate into that proxy object, and then uses `forwardingTargetForSelector:` for unknown selectors.

Should work by idea. 
But the real thing happens in `fetcherDidBeginFetching:`:
```
// PROBLEM STARTS HERE: this class is not dispatcher, it's TrustKit wrapper, and, maybe 
// subwrapped if you lucky who's using facebook's sdk as well... whatever
// the reason why it's crashing → GTMSessionFetcher expects there to be either parent 
// GTMSessionFetcher (who's sharing it's session) or GTMSessionFetcherSessionDelegateDispatcher

    GTMSessionFetcherSessionDelegateDispatcher *delegateDispatcher =
          [self delegateDispatcherForFetcher:fetcher];

// Sanity check that the fetcher's session is the delegate's shared session.
    NSURLSession *sharedSession = delegateDispatcher.session;
    NSURLSession *fetcherSession = fetcher.session;

    // so, we're asked for NSURLSession → and we got it, but clean from any GTM*** code! 
    if (sharedSession != nil && fetcherSession == sharedSession) {
       /* some code here */
       NSURLSessionTask *task = fetcher.sessionTask;
      if (task) {
        // CRASH AT THE NEXT LINE:
        [delegateDispatcher setFetcher:fetcher
                               forTask:task];
      }
    }
```

because of `setFetcher:forTask:` call is sent now to the instance of TrustKit swizzled object, which will try to forward it to the wrapped object, and will fail → Crash.

Reply:
> because of setFetcher:forTask: call is sent now to the instance of TrustKit swizzled object, which will try to forward it to the wrapped object, and will fail → Crash.

I'm still failing to see how what you describe has any swizzling.  You seemed to have described using a proxy delegate that forwards, but that doesn't require any swizzling.

And if they are just providing a proxy that forwards, isn't the -isKindOfClass: going to stop us from connecting things up?  i.e. - the proxy would come back `false` for that so we wouldn't return the delegate to have things relayed through.

Reply:
@thomasvl This ticket seems a little stale, but I've run into this issue as well with the New Relic iOS framework. We also swizzle NSURLSession and override the delegate object to inject a proxy. If I'm understanding this method correctly, `BOOL hasDispatcher = (fetcherDelegate != nil && fetcherDelegate != fetcher);`([GTMSessionFetcherService.m:384](https://github.com/google/gtm-session-fetcher/blob/master/Source/GTMSessionFetcherService.m#L384)) is asserting if `fetchDelegate` is not the `fetcher`, then it must be a `GTMSessionFetcherSessionDelegateDispatcher`. However, when a framework which instruments NSURLSession is introduced, proxying the delegate, this assumption is invalid. This results in the method `-setFetcher:forTask` potentially being called on a `GTMSessionFetcher` object. 

This is the issue I am currently experiencing. The proxied object is, in actuality, the fetcher, but because the check only compares the memory address, the memory address of `fetcherDelegate` will never match the memory address of `fetcher`. This results in the proxy object wrapping `fetcher` to be treated like a `GTMSessionFetcherSessionDelegateDispatcher` object. This results in -setFetcher:forTask to be called on the `GTMSessionFetcher` object.

I don't have full context, but using an 'isKindOfClass' maybe a better solution than a direct address check, unless the concern is there may be multiple unique instances of the GTMSessionFetcher object, but, if this was the case the logic flow would result in the same problem. 

I can't speak for geliso, but our proxy object is smart enough to verify 'isKindOfClass' against the wrapped delegate, and forward methods to that proxied object. So doing a 'isKindOfClass' check would perform as expected.

Reply:
What you describe makes some sense.  But this PR should break things then, as for the setup you describe, the new `maybeHasDispatcher` would be `False` and we'd then hit the `return nil` case, which doesn't seem what would be desired.

So it sounds like the best bet might be to just do the class check hoping proxy code "gets it right", and that's it.

@mwyman does that sound reasonable to you?

Reply:
@thomasvl @mwyman looks like it is exactly desired behaviour in this case. So, swizzled delegate won't be used as gtm session internals.

Reply:
I think the class check in place of the pointer value check sounds reasonable.